src/gt_frontend/py_to_gtscript.py

```diff
@@ -0,0 +1,229 @@
+import ast
+import enum
+import inspect
+import sys
+import typing
+
+import typing_inspect
+
+import eve.types
+import gtc.common
+from eve import UIDGenerator
+
+from . import ast_node_matcher as anm
+from . import gtscript_ast
+from .ast_node_matcher import Capture
+
+
+class PyToGTScript:
+    @staticmethod
+    def _all_subclasses(typ, *, module=None):
+        """
+            Return all subclasses of a given type. The type must be one of
+             - GT4PyAstNode (returns all subclasses of the given class)
+             - Union (return the subclasses of the united)
+             - ForwardRef (resolve the reference given the specified module and return its subclasses)
+             - built-in python type: str, int, type(None) (return as is)
+        """
+        if inspect.isclass(typ) and issubclass(typ, gtscript_ast.GTScriptAstNode):
+            result = set()
+            result.add(typ)
+            result.update(typ.__subclasses__())
+            result.update(
+                [
+                    s
+                    for c in typ.__subclasses__()
+                    for s in PyToGTScript._all_subclasses(c)
+                    if not inspect.isabstract(c)
+                ]
+            )
+            return result
+        elif inspect.isclass(typ) and typ in [
+            gtc.common.AssignmentKind,
+            gtc.common.UnaryOperator,
+            gtc.common.BinaryOperator,
+        ]:
+            # note: other types in gtc.common, e.g. gtc.common.DataType are not valid leaf nodes here as they
+            #  map to symbols in the gtscript ast and are resolved there
+            assert issubclass(typ, enum.Enum)
+            return set([typ])
+        elif typing_inspect.is_union_type(typ):
+            result = set()
+            for el_cls in typing_inspect.get_args(typ):
+                result.update(PyToGTScript._all_subclasses(el_cls, module=module))
+            return result
+        elif isinstance(typ, typing.ForwardRef):
+            type_name = typing_inspect.get_forward_arg(typ)
+            if not hasattr(module, type_name):
+                raise ValueError(
+                    "Reference to type `{}` in `ForwardRef` not found in module {}".format(
+                        type_name, module.__name__
+                    )
+                )
+            return PyToGTScript._all_subclasses(getattr(module, type_name), module=module)
+        elif typ in [
+            eve.types.StrictStr,
+            eve.types.StrictInt,
+            eve.types.StrictFloat,
+            str,
+            int,
+            float,
+            type(None),
+        ]:  # todo: enhance
+            return set([typ])
+
+        raise ValueError("Invalid field type {}".format(typ))
+
+    # template is a 1-to-1 mapping from context and python ast node to gt4py ast node. context is encoded in the field types
+    # all understood sementic is encoded in the structure
+    class Templates:
```

Comment:
Just a comment here: it's probably a matter of personal preference, but I think that renaming `Template` to `Pattern` and `Capture` to `MatchAs` would make the definitions more readable here.

Reply:
For the template/pattern I fully agree. For the `Capture` I have mixed feelings. On the one hand `Match` is the name that would resemble the wording in the regex module, on the other hand I always found this confusing as _to match_ and a _match_ suddenly mean different things. As capture is a well established name also used in the context of regexs (e.g. "capture groups") and _to capture_ blends well with natural language I will leave as is.

As a note, the patterns (to extract something) are also valid templates (to create something) since the mapping is bidirectional.